### PR TITLE
Formatcsv

### PR DESCRIPTION
**Pull request by spreeker**
seems to me this is a bunch of things together.   

CORS, 
csv improved loading on startup, 

csv added to endpoint.
could maybe create a branch with only those changes.

### CODE REVIEW

File: http_handlers.go [hunk 1]
```diff
@@ -26,26 +50,28 @@ func contextListRest(JWTConig jwtConfig, itemChan ItemsChannel, operations Group
...
 		groupByS, groupByFound := r.URL.Query()["groupby"]
+
 		if !groupByFound {
-			json.NewEncoder(w).Encode(items)
+			if query.ReturnFormat == "csv" {
+				writeCSV(items, w)
+			} else {
+				json.NewEncoder(w).Encode(items)
+			}
+			// force empty helps garbadge collection
```
Comment (Attumm) [inline http_handlers.go]: typo, garbage

File: main.go [hunk 2]
```diff
@@ -86,19 +92,28 @@ func main() {
...
+	mux.HandleFunc("/help/", helpRest)
+
+	if SETTINGS.Get("readonly") != "yes" {
+		mux.HandleFunc("/add/", addRest)
+		mux.HandleFunc("/rm/", rmRest)
+		mux.HandleFunc("/save/", saveRest)
+		mux.HandleFunc("/load/", loadRest)
+
+		mux.Handle("/", http.FileServer(http.Dir("./www2")))
+		mux.Handle("/dsm-search", http.FileServer(http.Dir("./www2")))
```
Comment (Attumm) [inline main.go]: this change is referencing www2 directory, that is not present.

File: http_handlers.go [hunk 3]
```diff
@@ -139,47 +209,21 @@ func loadRest(w http.ResponseWriter, r *http.Request) {
...
-		kSet[key] = true
-		LOOKUP[key] = append(LOOKUP[key], item)
-	}
-
-	keys := []string{}
-	for key := range kSet {
-		keys = append(keys, key)
-	}
+	// empty input save the memory
+	// TODO do not use ReadAll..but do it line by line
```
Comment (Attumm) [inline http_handlers.go]: TODO comment should be on line 210.

File: http_handlers.go [hunk 4]
```diff
@@ -124,10 +150,54 @@ func getStringFromIndex(data []byte, index int) string {
...
+	})
+
+	LOOKUP = make(map[string]Items)
+	kSet := make(map[string]bool)
+
+	for _, item := range ITEMS {
+		key := strings.ToLower(item.Buurtcode)
+		kSet[key] = true
+		LOOKUP[key] = append(LOOKUP[key], item)
+	}
```
Comment (Attumm) [inline http_handlers.go:167]: Let's create a "getIndex" that is created in the models.go file.
Then this code doesn't have to use "Buurtcode but instead will call getIndex.
```
sort.Slice(ITEMS, func(i, j int) bool {
		return ITEMS[i].getIndex() < ITEMS[j].getIndex()
	})
```

```
key := strings.ToLower(item.getIndex())
```
models.go
```
func (i item*) getIndex() string {
     return i.Buurtcode
}
```

File: operations.go [hunk 5]
```diff
@@ -266,7 +279,10 @@ func filteredEarlyExitSingle(items Items, column string, operations GroupedOpera
 	}
 	results := []string{}
 	for k := range filteredItemsSet {
-		results = append(results, k)
+		//empty keys crashes frontend.
+		if len(k) > 0 {
+			results = append(results, k)
+		}
```
Comment (Attumm) [inline operations.go:287]: Let's add comment here that this should be removed when the frontend doesn't break.

**Review by Attumm — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 6]
```diff
@@ -20,6 +20,8 @@ FROM scratch
 COPY --from=builder /app/main /app/main
 COPY --from=builder /app/www /www
 COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
+COPY extras/items.csv.gz /app/
 
-# Run the hello binary.
-ENTRYPOINT ["/app/main"]
+WORKDIR /app
+# Run the binary.
+ENTRYPOINT ["/app/main", "--csv", "items.csv.gz"]
```
Comment (Attumm) [inline Dockerfile]: what happens if items.csv.gz is not set?

File: csv.go [hunk 7]
```diff
@@ -1,10 +1,13 @@
 package main
 
 import (
+	"compress/gzip"
 	"encoding/json"
 	"errors"
 	"fmt"
 	"io"
+	//"io/ioutil"
```
Comment (Attumm) [inline csv.go]: Could be removed

**Comment by Attumm** [comment]
overall really neat.
Question what did you do some testing with GC with removing ITEMS, what is better setting nil, or creating a empty one?

I think using nil, is more need because it conveys to machine and reader that it's no longer needed.

File: Dockerfile (hunk 6 above)
Comment (spreeker) [reply Dockerfile]: prints a log line that file is missing.

File: http_handlers.go (hunk 1 above)
Comment (spreeker) [reply http_handlers.go]: fixed

File: http_handlers.go (hunk 3 above)
Comment (spreeker) [reply http_handlers.go]: done!

File: http_handlers.go (hunk 4 above)
Comment (spreeker) [reply http_handlers.go:167]: fixed.

File: operations.go (hunk 5 above)
Comment (spreeker) [reply operations.go:287]: applied

File: Dockerfile (hunk 6 above)
Comment (spreeker) [reply Dockerfile]: docker files cannot be build without items.csv I will comment that stuff out.

File: http_handlers.go [hunk 8]
```diff
@@ -16,7 +18,29 @@ import (
...
+// API list headers
+func setHeader(items Items, w http.ResponseWriter, query Query, queryTime int64) {
+
+	headerData := getHeaderData(items, query, queryTime)
+
+	for key, val := range headerData {
+		w.Header().Set(key, val)
+	}
+
+	if query.ReturnFormat == "csv" {
```
Comment (Attumm) [inline http_handlers.go:30]: Nice touch adding it to query. +1

File: http_handlers.go [hunk 9]
```diff
@@ -122,10 +150,53 @@ func getStringFromIndex(data []byte, index int) string {
...
+	STR_INDEX = []byte("\x00" + strings.Join(keys, "\x00") + "\x00")
+	INDEX = suffixarray.New(STR_INDEX)
+
+	fmt.Printf(WarningColorN, "sorted")
+}
+
+func writeCSV(items Items, w http.ResponseWriter) {
+	writer := csv.NewWriter(w)
+	for i := range items {
+		writer.Write(items[i].Row())
```
Comment (Attumm) [inline http_handlers.go:185]: Nice touch using Row.

**Pull request merged**